codalab/lib/bundle_util.py

```diff
@@ -354,3 +354,52 @@ def check_bundle_freezable(bundle):
             'Cannot freeze bundle %s(%s), bundle is not in a final state.'
             % (bundle.uuid, bundle.metadata.to_dict()["name"])
         )
+
+
+def get_bundle_state_details(bundle):
+    """
+    This helper returns information about the given bundle's status based on
+    the bundle's current `state` and `run_status`.
+
+    If the bundle is currently running, its `run_status` will be returned.
+    Otherwise, a description of the `state` is returned.
+    """
+    if 'state' not in bundle or 'bundle_type' not in bundle:
+        return ''
+
+    type = bundle['bundle_type']
+    state = bundle['state']
+
+    # see bundle_state.py for details on the State class
+    # see Bundle-Lifecycle.md for public bundle state documentation
+    state_details_by_type = {
+        'dataset': {
+            'created': 'Bundle has been created but its contents have not been uploaded yet.',
+            'uploading': 'Bundle contents are being uploaded.',
+            'ready': 'Bundle has finished uploading successfully, and is ready to be used for further runs.',
+            'failed': 'Bundle uploading failed.',
+        },
+        'make': {
+            'created': 'Bundle has been created but its contents have not yet been populated.',
+            'making': 'Bundle contents are being populated by copying its dependencies.',
+            'ready': 'Bundle contents have been successfully populated and is ready for further runs.',
+            'failed': 'Populating bundle contents failed.',
+        },
+        'run': {
+            'created': 'Bundle has been created but its contents have not been populated yet.',
+            'staged': 'Bundle’s dependencies are all ready. Waiting for the bundle to be assigned to a worker to be run.',
+            'starting': 'Bundle has been assigned to a worker and waiting for worker to start the bundle.',
+            'preparing': 'Waiting for worker to download bundle dependencies and Docker image to run the bundle.',
+            'finalizing': 'Bundle command has finished executing, cleaning up on the worker.',
+            'ready': 'Bundle command is finished executing successfully, and results have been uploaded to the server.',
+            'failed': 'Bundle has failed.',
+            'killed': 'Bundle was killed by the user. Bundle contents populated based on when the bundle was killed.',
+            'worker_offline': 'The worker where the bundle is running on is offline, and the worker might or might not come back online.',
+            'running': '',  # empty string fallback prevents crash in the unlikely event that `run_status` is unavailable.
+        },
+    }
+
+    # if the bundle is running, return the run status
```

Comment:
Let's add the staged status back...this should show up when we don't have enough resources to run a job (you might be able to do `cl run --request-gpus 10` or something to trigger).  This is pretty important to see show up.

Reply:
going to split this out (re: comment above)